lib/ae_mdw/activities.ex

```diff
@@ -152,8 +167,55 @@ defmodule AeMdw.Activities do
     end
   end
 
-  defp build_gens_stream(_state, _direction, _account_pk, _gen_scope, _cursor) do
-    []
+  defp build_gens_stream(gen_activities, direction) do
+    Stream.flat_map(gen_activities, fn [{height, _data} | _rest] = chunk ->
+      gen_events =
+        chunk
+        |> Enum.sort()
+        |> Enum.with_index()
+        |> Enum.map(fn {{^height, data}, local_idx} -> {{height, -1, local_idx}, data} end)
+
+      if direction == :forward do
+        gen_events
+      else
+        Enum.reverse(gen_events)
+      end
+    end)
+  end
+
+  defp build_int_transfers_stream(state, direction, account_pk, gen_scope, gen_cursor) do
+    @kinds
+    |> Enum.map(fn kind ->
+      key_boundary =
+        case gen_scope do
+          {first_gen, last_gen} ->
+            {
+              {account_pk, kind, {first_gen, -1}, @min_int},
+              {account_pk, kind, {last_gen, -1}, @max_int}
+            }
+
+          nil ->
+            {
+              {account_pk, kind, {@min_int, -1}, @min_int},
+              {account_pk, kind, {@max_int, -1}, @max_int}
+            }
+        end
+
+      cursor =
+        case gen_cursor do
+          nil -> nil
+          gen when direction == :forward -> {account_pk, kind, {gen, -1}, @min_int}
+          gen when direction == :backward -> {account_pk, kind, {gen, -1}, @max_int}
```

Comment:
Can be independent of `direction` as the the last element of the tuple is always -1.

Reply:
Done